lib/openssl/ssl.rb

```diff
@@ -168,6 +166,27 @@ def verify_certificate_identity(cert, hostname)
     end
     module_function :verify_certificate_identity
 
+    def verify_hostname(san, hostname)
+      san_parts = san.split(".")
+      return san == hostname if san_parts.size < 3
+
+      host_parts = hostname.split(".")
+      return false unless san_parts.size == host_parts.size
+      return false unless verify_domain_component(san_parts.shift, host_parts.shift)
+
+      san_parts.join(".") == host_parts.join(".")
+    end
+    module_function :verify_hostname
+
+    def verify_domain_component(san_component, domain_component)
```

Comment:
Should define the contract for this function, but looks like it does what I would expect (comparing to http://tools.ietf.org/html/rfc2818#section-3.1).